File: GLLara/GLLItemController.m
//
//  GLLItemController.m
//  GLLara
//
//  Created by Torsten Kammer on 17.09.12.
//  Copyright (c) 2012 Torsten Kammer. All rights reserved.
//

#import "GLLItemController.h"

#import "GLLItem.h"
#import "GLLBoneListController.h"
#import "GLLMeshListController.h"
#import "GLLSubItemController.h"
#import "NSArray+Map.h"
#import "GLLara-Swift.h"

@interface GLLItemController ()
// How many controllers for specific topics (as opposed to child items) there
// are. Two normally, three if the item has optional parts.
@property (nonatomic, readonly) NSUInteger fixedControllerCount;
@end

@implementation GLLItemController

- (id)initWithItem:(GLLItem *)item outlineView:(NSOutlineView *)outlineView parent:(id)parentController;
{
	if (!(self = [super init])) return nil;
	
	_outlineView = outlineView;
	
	self.item = item;
	self.parentController = parentController;
	self.meshListController = [[GLLMeshListController alloc] initWithItem:item parent:self];
	self.boneListController = [[GLLBoneListController alloc] initWithItem:item outlineView:outlineView parent:self];
    
    if (self.item.hasOptionalParts) {
        self.optionalPartsController = [[GLLOptionalPartController alloc] initWithItem:item parent:self];
    }
	
	[[NSNotificationCenter defaultCenter] addObserverForName:NSManagedObjectContextObjectsDidChangeNotification object:self.item.managedObjectContext queue:[NSOperationQueue mainQueue] usingBlock:^(NSNotification *notification){
		
		NSArray *removedItemControllers = [self.childrenControllers filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"item in %@", notification.userInfo[NSDeletedObjectsKey]]];
		[self.childrenControllers removeObjectsInArray:removedItemControllers];
		
		NSSet *addedItems = [notification.userInfo[NSInsertedObjectsKey] filteredSetUsingPredicate:[NSPredicate predicateWithFormat:@"entity.name == \"GLLItem\""]];
		NSArray *addedItemControllers = [addedItems map:^(GLLItem *item){
			if ([notification.userInfo[NSDeletedObjectsKey] containsObject:item]) return (id) nil; // Objects that were deleted again before this was called.
			if (item.parent != self.item) return (id) nil;
			return (id) [[GLLSubItemController alloc] initWithItem:item outlineView:self.outlineView parent:self];
		}];
		[self.childrenControllers addObjectsFromArray:addedItemControllers];
		[self.childrenControllers sortUsingDescriptors:@[ [NSSortDescriptor sortDescriptorWithKey:@"item.displayName" ascending:YES] ]];
		
		[self.outlineView reloadItem:self reloadChildren:YES];
	}];
	
	// Get initial child items
	NSFetchRequest *itemRequest = [NSFetchRequest fetchRequestWithEntityName:@"GLLItem"];
	itemRequest.sortDescriptors = @[ [NSSortDescriptor sortDescriptorWithKey:@"displayName" ascending:YES] ];
	itemRequest.predicate = [NSPredicate predicateWithFormat:@"parent == %@", self.item];
	
	self.childrenControllers = [[self.item.managedObjectContext executeFetchRequest:itemRequest error:NULL] mapMutable:^(GLLItem *item){
		return [[GLLSubItemController alloc] initWithItem:item outlineView:self.outlineView parent:self];
	}];
		
	return self;
}

- (id)representedObject
{
	return self.item;
}

- (NSArray *)allSelectableControllers
{
	NSMutableArray *result = [NSMutableArray arrayWithObject:self];
	[result addObjectsFromArray:self.meshListController.allSelectableControllers];
	[result addObjectsFromArray:self.boneListController.allSelectableControllers];
    if (self.optionalPartsController) {
        [result addObject:self.optionalPartsController];
    }
	return result;
}

- (NSUInteger)fixedControllerCount {
    return self.optionalPartsController ? 3 : 2;
}

#pragma mark - Outline View Data Source

- (id)outlineView:(NSOutlineView *)outlineView child:(NSInteger)index ofItem:(id)item
{
	switch (index)
	{
		case 0: return self.meshListController;
		case 1: return self.boneListController;
        case 2:
            if (self.optionalPartsController) {
                return self.optionalPartsController;
            } else {
                return self.childrenControllers[index - self.fixedControllerCount];
            }
		default: return self.childrenControllers[index - self.fixedControllerCount];
	}
}

- (id)outlineView:(NSOutlineView *)outlineView objectValueForTableColumn:(NSTableColumn *)tableColumn byItem:(id)item
{
	return self.item.displayName;
}

- (void)outlineView:(NSOutlineView *)outlineView setObjectValue:(id)object forTableColumn:(NSTableColumn *)tableColumn byItem:(id)item
{
	self.item.displayName = object;
}

- (BOOL)outlineView:(NSOutlineView *)outlineView isItemExpandable:(id)item
{
	return YES;
}

- (NSInteger)outlineView:(NSOutlineView *)outlineView numberOfChildrenOfItem:(id)item
{
	return self.fixedControllerCount + self.childrenControllers.count;
}

#pragma mark - Outline View Delegate

- (BOOL)outlineView:(NSOutlineView *)outlineView shouldEditTableColumn:(NSTableColumn *)tableColumn item:(id)item
{
	return YES;
}

@end

File: GLLara/GLLItemController.swift
//
//  GLLItemController.m
//  GLLara
//
//  Created by Torsten Kammer on 17.09.12.
//  Copyright (c) 2012 Torsten Kammer. All rights reserved.
//

import Cocoa

/*!
 * Source list controller for an item.
 */
@objc class GLLItemController: GLLItemListController {
    @objc convenience init(item: GLLItem, outlineView: NSOutlineView, parent: AnyObject) {
        self.init(item: item, outlineView: outlineView, parent: parent, showBones: true)
    }
    
    init(item: GLLItem, outlineView: NSOutlineView, parent: AnyObject, showBones: Bool) {
        super.init(managedObjectContext: item.managedObjectContext!, outlineView: outlineView, parentItem: item)
        
        meshListController = GLLMeshListController(item: item, parent: self)
        if showBones {
            boneListController = GLLBoneListController(item: item, outlineView: outlineView, parent: self)
        }
        if item.hasOptionalParts {
            optionalPartsController = GLLOptionalPartController(item: item, parent: self)
        }
    }
    
    @objc override var representedObject: AnyObject? {
        return item
    }
    
    @objc override var allSelectableControllers: [AnyObject] {
        var controllers: [AnyObject] = []
        if let meshListController {
            controllers.append(contentsOf: meshListController.allSelectableControllers)
        }
        if let boneListController {
            controllers.append(contentsOf: boneListController.allSelectableControllers)
        }
        if let optionalPartsController {
            controllers.append(optionalPartsController)
        }
        for childrenController in itemControllers {
            controllers.append(contentsOf: childrenController.allSelectableControllers)
        }
        return controllers
    }
    @objc weak var parentController: AnyObject?
    
    var meshListController: GLLMeshListController? = nil
    var boneListController: GLLBoneListController? = nil
    var optionalPartsController: GLLOptionalPartController? = nil
    
    override func outlineView(_ outlineView: NSOutlineView, numberOfChildrenOfItem item: Any?) -> Int {
        var result = 0
        if meshListController != nil {
            result += 1
        }
        if boneListController != nil {
            result += 1
        }
        if optionalPartsController != nil {
            result += 1
        }
        result += itemControllers.count
        return result
    }
    
    override func outlineView(_ outlineView: NSOutlineView, objectValueFor tableColumn: NSTableColumn?, byItem item: Any?) -> Any? {
        return self.item!.displayName
    }
    
    override func outlineView(_ outlineView: NSOutlineView, child index: Int, ofItem item: Any?) -> Any {
        var relativeIndex = index
        if let meshListController {
            if relativeIndex == 0 {
                return meshListController
            }
            relativeIndex -= 1
        }
        if let boneListController {
            if relativeIndex == 0 {
                return boneListController
            }
            relativeIndex -= 1
        }
        if let optionalPartsController {
            if relativeIndex == 0 {
                return optionalPartsController
            }
            relativeIndex -= 1
        }
        return itemControllers[relativeIndex]
    }
    
    func outlineView(_ outlineView: NSOutlineView, shouldEdit tableColumn: NSTableColumn?, item: Any) -> Bool {
        return true
    }
    
    override func outlineView(_ outlineView: NSOutlineView, isGroupItem item: Any) -> Bool {
        return false
    }
    
    override func outlineView(_ outlineView: NSOutlineView, shouldSelect tableColumn: NSTableColumn?) -> Bool {
        return true
    }
    
    func outlineView(_ outlineView: NSOutlineView, setObjectValue object: Any?, for tableColumn: NSTableColumn?, byItem some: Any?) {
        if let name = object as? String {
            self.item!.displayName = name
        }
    }
}
